Generate serialization of CurlProxySettings

donny-dont opened the pull request:
#### 38694a9a62607c8f9f4ad897d6c3257b57c2258a
<pre>
Generate serialization of CurlProxySettings
<a href="https://bugs.webkit.org/show_bug.cgi?id=272849">https://bugs.webkit.org/show_bug.cgi?id=272849</a>

Reviewed by Chris Dumez and Alex Christensen.

Move serialization of `CurlProxySettings` into a `.serialization.in`. file.

* Source/WebCore/platform/network/curl/CurlProxySettings.cpp:
* Source/WebCore/platform/network/curl/CurlProxySettings.h:
* Source/WebKit/Platform/Curl.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/curl/WebCoreArgumentCodersCurl.cpp: Removed.
* Source/WebKit/Shared/curl/WebCoreArgumentCodersCurl.serialization.in: Added.

Canonical link: <a href="https://commits.webkit.org/277705@main">https://commits.webkit.org/277705@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/587d2a7289fe089469265060952e3e04a078e7b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27549 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51291 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51025 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44402 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50642 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33485 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25070 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48919 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25237 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20634 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22717 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6393 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/44673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52929 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/23384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/19725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46825 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24649 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41936 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45739 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/10667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/25454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6876 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24372 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/network/curl/CurlProxySettings.cpp:
```diff
@@ -41,6 +41,14 @@ static constexpr uint16_t SocksProxyPort = 1080;
 
 static std::optional<String> createProxyUrl(const URL&);
 
+std::optional<CurlProxySettings> CurlProxySettings::tryMakeProxySettings(Mode mode, String&& url, String&& ignoreHosts)
```
This never returns std::nullopt it seems? Why return a std::optional?

donny-dont replied:
Oh I thought the serialization framework assumed thered be a `std::optional`. If that's not the case then I can do that.

cdumez replied:
Not 100% sure when using `CreateUsing` in the .in file. However, you can definitely stop using `CreateUsing` and use a constructor instead. And probably should if there is no failure case.

cdumez commented on Source/WebKit/Shared/curl/WebCoreArgumentCodersCurl.serialization.in:
```diff
@@ -0,0 +1,37 @@
+#if USE(CURL)
+
+[CreateUsing=tryMakeProxySettings] class WebCore::CurlProxySettings {
+    WebCore::CurlProxySettings::Mode mode();
```
The new code is a bit less efficient than the old one. Ideally we'd send a `std::variant<Mode, std::pair<String, String>>` or something similar so we don't have to IPC 3 things when the mode is not custom.

donny-dont replied:
Is that more efficient though? The `Mode` is an `uint8_t` and if there's no proxy then the `String`s should be empty.

cdumez replied:
Strings would be null but would still need to be serialized (as 2 booleans that are false).

cdumez replied:
Likely not a massive issue but that's the approach we've taken in many places and it would match the previous serialization logic more closely.

donny-dont replied:
Ok will adjust accordingly

cdumez commented on Source/WebCore/platform/network/curl/CurlProxySettings.h:
```diff
@@ -62,6 +62,10 @@ class CurlProxySettings {
     long authMethod() const { return m_authMethod; }
 
+    using PlatformDataVariant = std::variant<Mode, std::pair<URL, String>>;
+    WEBCORE_EXPORT PlatformDataVariant platformSerializationData() const;
```
I think the pattern in the rest of the code base is to call this `toIPCData()` and the variant `IPCData`.

cdumez commented on Source/WebCore/platform/network/curl/CurlProxySettings.h:
```diff
@@ -62,6 +62,10 @@ class CurlProxySettings {
 
+    using PlatformDataVariant = std::variant<Mode, std::pair<URL, String>>;
+    WEBCORE_EXPORT PlatformDataVariant platformSerializationData() const;
+    WEBCORE_EXPORT static CurlProxySettings makeProxySettings(PlatformDataVariant);
```
Should be `PlatformDataVariant&&` or `IPCData&&` really. No reason to copy the strings and refcount churn.

achristensen07 commented on Source/WebCore/platform/network/curl/CurlProxySettings.h:
```diff
@@ -62,6 +62,10 @@ class CurlProxySettings {
     void setAuthMethod(long);
     long authMethod() const { return m_authMethod; }
 
+    using PlatformDataVariant = std::variant<Mode, std::pair<URL, String>>;
```
This is fine, but it makes it possible to send Mode::Custom without a URL/String pair.  A further refinement would be something like this:

struct DefaultData { };
struct NoProxyData { };
struct CustomData {
    URL url;
    String ignoreHosts;
};
using PlatformDataVariant = std::variant<DefaultData, NoProxyData, CustomData>;

Then also use PlatformDataVariant for member storage if you really want to go to town.

cdumez commented on Source/WebCore/platform/network/curl/CurlProxySettings.cpp:
```diff
@@ -89,6 +89,26 @@ bool protocolIsInSocksFamily(const URL& url)
+CurlProxySettings CurlProxySettings::makeProxySettings(CurlProxySettings::PlatformDataVariant platformData)
+{
+    return WTF::switchOn(platformData,
+        [&] (Mode d) {
```
Please use a better name than `d` :)

cdumez commented on Source/WebCore/platform/network/curl/CurlProxySettings.cpp:
```diff
@@ -89,6 +89,26 @@ bool protocolIsInSocksFamily(const URL& url)
+        [&] (Mode d) {
+            return CurlProxySettings { d };
+        },
+        [&] (std::pair<URL, String>& d) {
```
Please use a better name than `d` :)

cdumez reviewed (approved):
LGTM with nits.

donny-dont committed:
Generate serialization of CurlProxySettings
https://bugs.webkit.org/show_bug.cgi?id=272849

Reviewed by Chris Dumez and Alex Christensen.

Move serialization of `CurlProxySettings` into a `.serialization.in`. file.

* Source/WebCore/platform/network/curl/CurlProxySettings.cpp:
* Source/WebCore/platform/network/curl/CurlProxySettings.h:
* Source/WebKit/Platform/Curl.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/curl/WebCoreArgumentCodersCurl.cpp: Removed.
* Source/WebKit/Shared/curl/WebCoreArgumentCodersCurl.serialization.in: Added.

Canonical link: https://commits.webkit.org/277705@main
